### PR TITLE
erts: Fix a race in persistent_term:get/1,2

### LINKED ISSUE #5908: persistent_term:get could return random term when racing happens
**Describe the bug**
In very rare case, we see persistent_term:get/2 returns a random value that has never been set.

For example, we only have 2 write operations for key 'some_key': persistent_term:put('some_key', true) and persistent_term:erase('some_key'), but persistent_term:get('some_key', false) could return an integer or a reference or anything.

Our theory is, persistent_term_get_2 does not have locks, and  *hash_table can be changed during lookup and get_bucket operations.
https://github.com/erlang/otp/blob/master/erts/emulator/beam/erl_bif_persistent.c#L490

Basically, if some persistent_term write operations (erase, put, which changes the erts_atomic_read_nob(&the_hash_table)) happened when doing persistent_term_get_2, *hash_table is possibly freed and filled by anything.

If it also happens that get_bucket(hash_table, entry_index) is a tuple of {_, RandomTerm}, then persistent_term_get_2 will return RandomTerm as result.

**To Reproduce**
This is a really rare racing condition. We can only reproduce that erts_atomic_read_nob(&the_hash_table) can be changed to something different with hash_table while doing lookup and get_bucket operation
see https://github.com/potatosalad/otp/commit/7b460431c53a5715d4ca6d517ccb876470fb7b81

**Expected behavior**
persistent_term:get should never return a value that has never been set.

This might be achievable by adding a read write lock to hash_table: Basically, erts_atomic_read_nob(&the_hash_table) should not be changed when there is an on-going read operation.

If this hurts the performance a lot, we could at least add a lookup_term function which does an atomic operation instead of two steps: entry_index = lookup(hash_table, key); get_bucket(hash_table, entry_index);

**Affected versions**
erts-12.1.2

### PR DESCRIPTION
**Pull request by jhogberg**
Fixes #5908

### CODE REVIEW

File: erts/emulator/beam/erl_bif_persistent.c [hunk 1]
```diff
@@ -953,19 +960,25 @@ cleanup_trap_data(Binary *bp)
...
 
     while (1) {
         term = get_bucket(hash_table, idx & mask);
-        if (is_nil(term) || EQ(key, (tuple_val(term))[1]))
-            break;
+
+        if (is_nil(term) || EQ(key, (tuple_val(term))[1])) {
+            if (bucket) {
+                *bucket = term;
+            }
```
Comment (sverker) [inline erts/emulator/beam/erl_bif_persistent.c:975]: `bucket` argument is never NULL to `lookup()` what I can see

File: erts/emulator/beam/erl_bif_persistent.c [hunk 2]
```diff
@@ -359,7 +360,7 @@ BIF_RETTYPE persistent_term_put_2(BIF_ALIST_2)
     ctx->key = BIF_ARG_1;
     ctx->term = BIF_ARG_2;
 
-    ctx->entry_index = lookup(ctx->hash_table, ctx->key);
+    ctx->entry_index = lookup(ctx->hash_table, ctx->key, &ctx->old_bucket);
```
Comment (sverker) [inline erts/emulator/beam/erl_bif_persistent.c:363]: The first `get_bucket()` call in `persistent_term_put_2` can be replaced with `ctx->old_bucket`:

`if (is_nil(get_bucket(ctx->hash_table, ctx->entry_index))) {
`

File: erts/emulator/beam/erl_bif_persistent.c [hunk 3]
```diff
@@ -122,6 +122,7 @@ typedef struct {
     HashTable* hash_table;
     Eterm heap[3];
     Eterm tuple;
+    Eterm old_bucket;
```
Comment (sverker) [inline erts/emulator/beam/erl_bif_persistent.c:125]: `old_bucket` does not need to be in ErtsPersistentTermPut2Context as it's not used over traps, but it doesn't do any harm either so...

**Review by sverker — APPROVED** [review]

File: erts/emulator/beam/erl_bif_persistent.c (hunk 1 above)
Comment (jhogberg) [reply erts/emulator/beam/erl_bif_persistent.c]: Thanks, fixed. :)

File: erts/emulator/beam/erl_bif_persistent.c (hunk 2 above)
Comment (jhogberg) [reply erts/emulator/beam/erl_bif_persistent.c]: Good catch, thanks!

File: erts/emulator/beam/erl_bif_persistent.c (hunk 3 above)
Comment (jhogberg) [reply erts/emulator/beam/erl_bif_persistent.c]: Good point, fixed. :)

**Review by max-au — APPROVED** [review]

**Pull request merged**